lib/mgmt_msg_native.h

```diff
@@ -115,6 +261,118 @@ extern int vmgmt_msg_native_send_error(struct msg_conn *conn,
+					      sizeof(msg_type) + (var_len),    \
+					      mem_type);                       \
+	})
+
+/**
+ * mgmt_msg_free_native_msg() - Free a native msg.
+ * @msg - pointer to message allocated by mgmt_msg_create_native_msg().
+ */
+#define mgmt_msg_native_free_msg(msg) darr_free(msg)
+
+/**
+ * mgmt_msg_native_get_msg_len() - Get the total length of the msg.
+ * @msg: the native message.
+ *
+ * Return: the total length of the message, fixed + variable length.
+ */
+#define mgmt_msg_native_get_msg_len(msg) (darr_len((uint8_t *)(msg)))
+
+/**
+ * mgmt_msg_native_append() - Append data to the end of the msg.
+ * @msg: (IN/OUT) Pointer to the native message, variable may be updated.
+ * @data: data to append.
+ * @len: length of data to append.
+ *
+ * Append @data of length @len to the native message @msg.
+ *
+ * NOTE: Be aware @msg pointer may change as a result of reallocating the
+ * message to fit the new data. Any other pointers into the old message should
+ * be discarded.
+ *
+ * Return: a pointer to the newly appended data.
+ */
+#define mgmt_msg_native_append(msg, data, len)                                 \
+	memcpy(darr_append(*mgmt_msg_get_native_darrp(msg), len), data, len)
+
+/**
+ * mgmt_msg_native_send_msg(msg, short_circuit_ok) - Send a native msg.
+ * @conn: the mgmt_msg connection.
+ * @msg: the native message.
+ * @short_circuit_ok: True if short-circuit sending is required.
+ *
+ * Return: The error return value of msg_conn_send_msg().
+ */
+#define mgmt_msg_native_send_msg(conn, msg, short_circuit_ok)                  \
+	msg_conn_send_msg(conn, MGMT_MSG_VERSION_NATIVE, msg,                  \
+			  mgmt_msg_native_get_msg_len(msg), NULL,              \
+			  short_circuit_ok)
+
+/**
+ * mgmt_msg_native_get_darrp() - Return a ptr to the dynamic array ptr.
+ * @msg: Pointer to the native message.
+ *
+ * NOTE: Most users can simply use mgmt_msg_native_append() instead of this.
+ *
+ * This function obtains a pointer to the dynamic byte array for this message,
+ * this array actually includes the message header if one is going to look at
+ * the length value. With that in mind any of the `darr_*()` functions/API may
+ * be used to manipulate the variable data at the end of the message.
+ *
+ * NOTE: The pointer returned is actually a pointer to the message pointer
+ * passed in to this function. This pointer to pointer is required so that
+ * realloc can be done inside the darr API.
+ *
+ * NOTE: If reallocs happen the original passed in pointer will be updated;
+ * however, any other pointers into the message will become invalid and so they
+ * should always be discarded after using the returned value.
+ *
+ * Example:
+ *
+ *     void append_metric_path(struct mgmt_msg_my_msg *msg)
+ *     {
+ *          char *xpath = msg->xpath;	// pointer into message
+ *          uint8_t **darp;
+ *
+ *          darrp = mgmt_msg_native_get_darrp(msg);
+ *          darr_in_strcat(*darrp, "/metric");
```

Comment:
Basically the entire darr API are macros, and all the reallocating API modify the passed in pointer variable directly. Just like `XFREE(mtype, foo)` NULL's `foo` etc.

Reply:
Because darr_in_strcat is a macro not a function it's operating directly on the passed in code which leads to:
```
	darr_ensure_cap_mt(D, ...) -> _darr_resize_mt((D), ...)
...
#define _darr_resize_mt(A, C, MT)
	({ (A) = __darr_resize(A, C, _darr_esize(A), MT); })
```
So:
```
	*(&msg) = __darr_resize(*(&msg), ...)
```